src/test/java/com/google/roomies/database/FirebaseDatabaseTest.java

```diff
@@ -0,0 +1,100 @@
+package com.google.roomies.database;
+
+import static com.google.roomies.ListingConstants.LISTING_COLLECTION_NAME;
+import static com.google.roomies.ListingRequestParameterNames.DESCRIPTION;
+import static com.google.roomies.ListingRequestParameterNames.END_DATE;
+import static com.google.roomies.ListingRequestParameterNames.LISTING_PRICE;
+import static com.google.roomies.ListingRequestParameterNames.LEASE_TYPE;
+import static com.google.roomies.ListingRequestParameterNames.NUM_BATHROOMS;
+import static com.google.roomies.ListingRequestParameterNames.NUM_ROOMS;
+import static com.google.roomies.ListingRequestParameterNames.NUM_SHARED;
+import static com.google.roomies.ListingRequestParameterNames.NUM_SINGLES;
+import static com.google.roomies.ListingRequestParameterNames.SHARED_ROOM_PRICE;
+import static com.google.roomies.ListingRequestParameterNames.SINGLE_ROOM_PRICE;
+import static com.google.roomies.ListingRequestParameterNames.START_DATE;
+import static com.google.roomies.ListingRequestParameterNames.TIMESTAMP;
+import static com.google.roomies.ListingRequestParameterNames.TITLE;
+import static com.google.roomies.ProjectConstants.PROJECT_ID;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.mockito.Mockito.when;
+import static org.mockito.Mockito.verify;
+
+import com.google.api.core.ApiFuture;
+import com.google.api.core.ApiFutures;
+import com.google.cloud.firestore.CollectionReference;
+import com.google.cloud.firestore.DocumentReference;
+import com.google.cloud.firestore.DocumentSnapshot;
+import com.google.cloud.firestore.FieldValue;
+import com.google.cloud.firestore.Firestore;
+import com.google.cloud.firestore.FirestoreOptions.Builder;
+import com.google.cloud.firestore.Query;
+import com.google.cloud.firestore.QueryDocumentSnapshot;
+import com.google.cloud.firestore.QuerySnapshot;
+import com.google.cloud.firestore.WriteResult;
+import com.google.cloud.Timestamp;
+import com.google.common.collect.ImmutableMap;
+import com.google.roomies.Listing;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.io.StringWriter;
+import java.text.ParseException;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.mockito.MockitoAnnotations;
+
+@RunWith(JUnit4.class)
+public class FirebaseDatabaseTest {
```

Comment:
I don't want to block this PR on it, but this test should really test that all of the NoSQLDatabase interface methods are implemented correctly. That would have ideally happened at the same time FirestoreDatabase was initially added.

Reply:
Added tests for the other interface methods